compiler/il/OMRNode.cpp

```diff
@@ -2472,17 +2488,37 @@ OMR::Node::computeIsCollectedReference()
             // only problem is: we might mark an aladd/aiadd on a temp whose value is null as
             // an internal pointer, thus creating a temp, pinning array and internal pointer
             // map for it.
-            return self() == curNode && curNode->getAddress() == 0;
+
+            // Preserve the existing logic which is as follows:
+            // true iff not under aladd and null,
+            // if non-null, always false.
+            // if under aladd, always false.
+
+            // non-null constant return false.
+            if (curNode->getAddress() != 0)
+               return TR_no;
+            else
+               {
+               // null constant under aladd, return false.
+               if (self() != curNode)
+                  return TR_no;
+               else
+                  // Null constant under ternary (i.e. we reached here via recursive calls), return maybe
+                  // to indicate need to check the other child.
+                  // Otherwise return yes for top level Null const nodes, to preserve previous logic.
+                  return (toplevel ? TR_yes : TR_maybe);
```

Comment:
Shouldn't this be a constant maybe? If the top level says maybe the result doesn't have to be collected does it?

Reply:
Under previous logic, we had:
```
case TR::aconst:
    // aconst null can either be collected or uncollected. The collectedness of aconst
    // null can be determined by usedef analysis, which is too expensive for a very
    // basic query. Treating aconst null as a collected reference is always safe. The
    // only problem is: we might mark an aladd/aiadd on a temp whose value is null as
    // an internal pointer, thus creating a temp, pinning array and internal pointer
    // map for it.
    return self() == curNode && curNode->getAddress() == 0;
```
This says for Null aconst node, and self() == curNode meaning the aconst node is not under `aladd/aiadd`, we return true. 
In the new logic, if `toplevel ` is true, it means the aconst node is not under a ternary; in this case, I thought we should preserve the same logic as before and return true.

Reply:
I'm not sure this will be correct in all cases - surely we have to look at the children to know whether the overall result needs to be collected? if there is a large tree with many ternary operations and at least one is collected we need the result to be collected otherwise not. How can you know the situation without checking down to the leaves?

Reply:
Only when we recursively call this API on ternary value child is `toplevel` passed in as false - so if `toplevel` is false, we know this current node is a child of ternary (as this is a recursive call) and we return TR_maybe, and the caller will then do another recursive call on the other child and also passing in `toplevel` false. 

Only when user code calls this API, `toplevel` is passed in as true, so we know we are not called recursively therefore receiver node is not the value child of ternary - so if we end up here and `toplevel` is true, we can return TR_yes as previously implemented.

Reply:
I still don't buy it - you could have 5 levels of aternary and all the leaves are aconst null except one - you must visit that one to know what the result should be.

Reply:
After discussing this with Andrew, we believe this change should be able to handle nested ternary tree structure with multiple `aconst null` as leaf nodes. However, if a ternary node is referenced multiple times in a very deep tree structure, we would have to walking that same ternary node trees repeatedly as there is no indicator whether a node had already been processed. We discussed a solution where we use 2 checklists to record the result for ternary nodes that have been processed.

Reply:
This toplevel thing still seems to bei n the code - it should disappear now you have the checklists

Reply:
I think `toplevel` flag is still needed to differentiate whether a particular `aconst null` node is a root node or a leaf node under a ternary node, when it is first encountered. Since this call can be recursive, if we end up at this line, we wouldn't have known if the receiver node is a root node or leaf node under ternary, so we use this flag to tell us.
If it's a leaf node, we return `TR_yes`, otherwise return `TR_maybe` for lead node under ternary. For subsequent encounters, we check the nodelists and retrieve the previous result directly.